File: python/publisher.py
#!/usr/bin/env python3
# Copyright (c) 2025 Ziqi Fan
# SPDX-License-Identifier: Apache-2.0

"""
Publish cmd_vel messages to ROS2 using Zenoh
"""

import argparse
import time

import zenoh

from msg import Twist, Vector3


def main():
    parser = argparse.ArgumentParser(description="Zenoh cmd_vel publisher")
    parser.add_argument(
        "--bridge",
        "-b",
        type=str,
        required=True,
        help="Bridge address, e.g.: localhost:7447",
    )
    parser.add_argument(
        "--linear",
        "-l",
        type=float,
        default=0.5,
        help="Linear velocity (default: %(default)s)",
    )
    parser.add_argument(
        "--angular",
        "-a",
        type=float,
        default=0.2,
        help="Angular velocity (default: %(default)s)",
    )
    args = parser.parse_args()

    config = zenoh.Config()
    config.insert_json5("connect/endpoints", f'["tcp/{args.bridge}"]')

    session = zenoh.open(config)
    publisher = session.declare_publisher("cmd_vel")

    print(f"Zenoh cmd_vel publisher started")  # noqa: F541
    print(f"  Connection: tcp/{args.bridge}")
    print(f"  Topic: cmd_vel -> ROS2 /cmd_vel")  # noqa: F541
    print(f"  Velocity: linear.x={args.linear}, angular.z={args.angular}")
    print()

    try:
        while True:
            twist = Twist(
                linear=Vector3(x=args.linear, y=0.0, z=0.0),
                angular=Vector3(x=0.0, y=0.0, z=args.angular),
            )

            publisher.put(twist.serialize())
            print(f"Published: linear.x={args.linear}, angular.z={args.angular}")
            time.sleep(1.0)

    except KeyboardInterrupt:
        print("\nStopping")
    finally:
        session.close()


if __name__ == "__main__":
    main()

File: rust/src/bin/publisher.rs
// Copyright (c) 2025 Ziqi Fan
// SPDX-License-Identifier: Apache-2.0

use std::env;
use std::thread;
use std::time::Duration;
use zenoh::{Config, Wait};
use zenoh_ros2dds_example::cdr;
use zenoh_ros2dds_example::msg::{Twist, Vector3};

fn main() {
    let args: Vec<String> = env::args().collect();

    let bridge_addr = if args.len() > 1 {
        args[1].clone()
    } else {
        eprintln!("Error: Bridge address must be specified");
        eprintln!("Usage: {} <bridge_address> [linear_x] [angular_z]", args[0]);
        eprintln!("Example: {} localhost:7447 0.5 0.2", args[0]);
        return;
    };

    let linear_x: f64 = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(0.5);
    let angular_z: f64 = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(0.2);

    let mut config = Config::default();
    config
        .insert_json5("connect/endpoints", &format!(r#"["tcp/{}"]"#, bridge_addr))
        .expect("Failed to configure connection endpoints");

    let session = match zenoh::open(config).wait() {
        Ok(s) => s,
        Err(e) => {
            eprintln!("Connection failed: {}", e);
            eprintln!("Usage: {} [bridge_address] [linear_x] [angular_z]", args[0]);
            eprintln!("Example: {} 192.168.1.100:7447 0.5 0.2", args[0]);
            return;
        }
    };

    let topic = "cmd_vel";
    let publisher = match session.declare_publisher(topic).wait() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("Failed to create publisher: {}", e);
            return;
        }
    };

    println!("Zenoh cmd_vel publisher started");
    println!("  Connection: tcp/{}", bridge_addr);
    println!("  Topic: {} -> ROS2 /cmd_vel", topic);
    println!("  Velocity: linear.x={}, angular.z={}", linear_x, angular_z);
    println!();

    loop {
        let twist = Twist {
            linear: Vector3 {
                x: linear_x,
                y: 0.0,
                z: 0.0,
            },
            angular: Vector3 {
                x: 0.0,
                y: 0.0,
                z: angular_z,
            },
        };
        let payload = cdr::serialize(&twist).unwrap();

        if let Err(e) = publisher.put(payload).wait() {
            eprintln!("Publish error: {}", e);
        } else {
            println!("Published: linear.x={}, angular.z={}", linear_x, angular_z);
        }

        thread::sleep(Duration::from_secs(1));
    }
}
